middleware/hosts/setup.go

```diff
@@ -0,0 +1,88 @@
+package hosts
+
+import (
+	"log"
+	"os"
+	"path"
+
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/middleware"
+
+	"github.com/mholt/caddy"
+)
+
+func init() {
+	caddy.RegisterPlugin("hosts", caddy.Plugin{
+		ServerType: "dns",
+		Action:     setup,
+	})
+}
+
+func setup(c *caddy.Controller) error {
+	h, err := hostsParse(c)
+	if err != nil {
+		return middleware.Error("hosts", err)
+	}
+
+	dnsserver.GetConfig(c).AddMiddleware(func(next middleware.Handler) middleware.Handler {
+		h.Next = next
+		return h
+	})
+
+	return nil
+}
+
+func hostsParse(c *caddy.Controller) (Hosts, error) {
+	var h = Hosts{
+		Hostsfile: &Hostsfile{path: "/etc/hosts"},
+	}
+	defer h.ReadHosts()
+
+	config := dnsserver.GetConfig(c)
+
+	for c.Next() {
+		if c.Val() == "hosts" { // hosts [FILE] [ZONES...]
+			args := c.RemainingArgs()
+			if len(args) == 0 {
```

Comment:
Yeah, this does look off. I think the cause of your issue above may be related to the changes I made to add the Fallthrough config. I'll look into this tonight and add more tests.